src/Microsoft.OData.Edm/Csdl/CsdlReader.Json.cs

```diff
@@ -0,0 +1,201 @@
+    /// Provides CSDL-JSON parsing services for EDM models.
+    /// </summary>
+    public partial class CsdlReader
+    {
+        /// <summary>
+        /// Tries parsing the given CSDL-JSON artifact for an IEdmModel.
+        /// </summary>
+        /// <param name="reader">The given JSON reader containing the CSDL artifact.</param>
+        /// <param name="model">The model generated by parsing</param>
+        /// <param name="errors">Errors reported while parsing.</param>
+        /// <returns>Success of the parse operation.</returns>
+        public static bool TryParse(ref Utf8JsonReader reader, out IEdmModel model, out IEnumerable<EdmError> errors)
+        {
+            return TryParse(ref reader, CsdlJsonReaderSettings.Default, out model, out errors);
+        }
+
+        /// <summary>
+        /// Tries parsing the given CSDL-JSON artifact for an IEdmModel.
+        /// </summary>
+        /// <param name="reader">The given JSON reader containing the CSDL artifact.</param>
+        /// <param name="settings">CSDL-JSON reader settings for current parser.</param>
+        /// <param name="model">The model generated by parsing</param>
+        /// <param name="errors">Errors reported while parsing.</param>
+        /// <returns>Success of the parse operation.</returns>
+        public static bool TryParse(ref Utf8JsonReader reader, CsdlJsonReaderSettings options, out IEdmModel model, out IEnumerable<EdmError> errors)
+        {
+            EdmUtil.CheckArgumentNull(options, nameof(options));
+
+            model = null;
+
+            JsonParserContext context = new JsonParserContext(options, source: null);
+
+            CsdlModel mainModel = CsdlJsonParser.ParseCsdlDocument(ref reader, context);
+
+            if (mainModel != null && !context.HasIntolerableError())
+            {
+                Debug.Assert(mainModel.CsdlVersion != null, "csdlVersion != null");
+
+                List<CsdlModel> referencedAstModels = LoadReferencedCsdl(mainModel, context);
+
+                if (!context.HasIntolerableError())
+                {
+                    CsdlSemanticsModel tmp = new CsdlSemanticsModel(mainModel,
+                        new CsdlSemanticsDirectValueAnnotationsManager(),
+                        referencedAstModels,
+                        options.IncludeDefaultVocabularies);
+
+                    // add more referenced IEdmModels in addition to the above loaded CsdlModels.
+                    if (options.ReferencedModels != null)
+                    {
+                        tmp.AddToReferencedModels(options.ReferencedModels);
+                    }
+
+                    model = tmp;
+                    model.SetEdmxVersion(mainModel.CsdlVersion);
+                    Version edmVersion;
+                    if (CsdlConstants.EdmxToEdmVersions.TryGetValue(mainModel.CsdlVersion, out edmVersion))
+                    {
+                        model.SetEdmVersion(edmVersion);
+                    }

```

Comment:
if a referenced model has an intolerable error, should we still load the main model (with, perhaps, some unresolved references?)

Reply:
Yes, in current pattern , We still load the main model. 
for the referenced types, only output "unresolved element" if such element is used outside.
besides, that's the XML reader pattern.

---
In reply to: [429391404](https://github.com/OData/odata.net/pull/1775#discussion_r429391404) [](ancestors = 429391404)